### PR TITLE
misc bug fixes + llvm cache changes 

### PR DESCRIPTION
**Pull request by alexbaden**

### CODE REVIEW

**Comment by ienkovich** [comment]
`TableFunctions.BasicProjection` hangs for me on this branch.

**Comment by alexbaden** [comment]
Table Functions test hanging should be resolved. Bad rebase left two attempts at the same mutex.

File: QueryEngine/Execute.cpp [hunk 1]
```diff
@@ -482,9 +494,9 @@ StringDictionaryProxy* RowSetMemoryOwner::getOrAddStringDictProxy(
         with_generation ? string_dictionary_generations_.getGeneration(dict_id) : -1;
     return addStringDict(dd->stringDict, dict_id, generation);
   }
-  CHECK_EQ(0, dict_id);
+  CHECK_EQ(dict_id, DictRef::literalsDictId);
   if (!lit_str_dict_proxy_) {
-    DictRef literal_dict_ref{DictRef::InvalidDictRef()};
+    DictRef literal_dict_ref(DictRef::invalidDbId, DictRef::literalsDictId);
```
Comment (alexbaden) [inline QueryEngine/Execute.cpp:499]: @ienkovich Can you confirm the dict ID + schema ID introduced in https://github.com/intel-ai/omniscidb/commit/3fb06a27e5cf44cdeea7ba3c48f5493286495a64 can never be 0? If so, this change should be ok. 

Separately we can probably remove the dict id from dict ref struct.

Comment (ienkovich) [reply QueryEngine/Execute.cpp:499]: 0 dict_id is reserved for `TRANSIENT_DICT_ID`. We don't actually enforce schema ID to be non-zero, we used zero schema ID as a sign to use OmniSci storage. Now it should be unused, but we have no checks preventing users from using it.

Comment (alexbaden) [reply QueryEngine/Execute.cpp:499]: I think here that is one and the same -- the transient dictionary holds literals for certain case statement projections I believe.

**Review by ienkovich — APPROVED** [review]

**Comment by kurapov-peter** [comment]
The new caching broke most of gpu tests for me, i.e.:
```
C  0x0000000000000061
C  [ArrowBasedExecuteTest+0x1895501]  GpuDeviceCompilationContext::~GpuDeviceCompilationContext()+0xa9
C  [ArrowBasedExecuteTest+0x187c59c]  std::default_delete<GpuDeviceCompilationContext>::operator()(GpuDeviceCompilationContext*) const+0x22
C  [ArrowBasedExecuteTest+0x1876450]  std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >::~unique_ptr()+0x52
C  [ArrowBasedExecuteTest+0x1889063]  void std::_Destroy<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> > >(std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*)+0x18
C  [ArrowBasedExecuteTest+0x1884900]  void std::_Destroy_aux<false>::__destroy<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*>(std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*, std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*)+0x2e
C  [ArrowBasedExecuteTest+0x187f9b6]  void std::_Destroy<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*>(std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*, std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*)+0x23
C  [ArrowBasedExecuteTest+0x1879ec5]  void std::_Destroy<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*, std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> > >(std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*, std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >*, std::allocator<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> > >&)+0x27
C  [ArrowBasedExecuteTest+0x1873fc7]  std::vector<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> >, std::allocator<std::unique_ptr<GpuDeviceCompilationContext, std::default_delete<GpuDeviceCompilationContext> > > >::~vector()+0x35
C  [ArrowBasedExecuteTest+0x1892ada]  GpuCompilationContext::~GpuCompilationContext()+0x2a
C  [ArrowBasedExecuteTest+0x1892f67]  void __gnu_cxx::new_allocator<GpuCompilationContext>::destroy<GpuCompilationContext>(GpuCompilationContext*)+0x23
C  [ArrowBasedExecuteTest+0x1892ef3]  void std::allocator_traits<std::allocator<GpuCompilationContext> >::destroy<GpuCompilationContext>(std::allocator<GpuCompilationContext>&, GpuCompilationContext*)+0x23
C  [ArrowBasedExecuteTest+0x1892c6f]  std::_Sp_counted_ptr_inplace<GpuCompilationContext, std::allocator<GpuCompilationContext>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()+0x37
C  [ArrowBasedExecuteTest+0x1464ce2]  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+0x42
C  [ArrowBasedExecuteTest+0x145a0bb]  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count()+0x27
C  [ArrowBasedExecuteTest+0x166b8f2]  std::__shared_ptr<GpuCompilationContext, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr()+0x1c
C  [ArrowBasedExecuteTest+0x166b90e]  std::shared_ptr<GpuCompilationContext>::~shared_ptr()+0x18
C  [ArrowBasedExecuteTest+0x16cd8b6]  std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >::~pair()+0x1c
C  [ArrowBasedExecuteTest+0x16cd8e2]  void __gnu_cxx::new_allocator<std::_List_node<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > >::destroy<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > >(std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >*)+0x1c
C  [ArrowBasedExecuteTest+0x16bb333]  void std::allocator_traits<std::allocator<std::_List_node<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > > >::destroy<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > >(std::allocator<std::_List_node<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > >&, std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >*)+0x23
C  [ArrowBasedExecuteTest+0x16a4caa]  std::__cxx11::_List_base<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >, std::allocator<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > >::_M_clear()+0x62
C  [ArrowBasedExecuteTest+0x168ebc8]  std::__cxx11::_List_base<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >, std::allocator<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > >::~_List_base()+0x18
C  [ArrowBasedExecuteTest+0x1677af2]  std::__cxx11::list<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> >, std::allocator<std::pair<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext> > > >::~list()+0x18
C  [ArrowBasedExecuteTest+0x1677b3a]  LruCache<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<GpuCompilationContext>, boost::hash<std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > >::~LruCache()+0x28
C  [ArrowBasedExecuteTest+0x16f015e]  CodeCacheAccessor<GpuCompilationContext>::~CodeCacheAccessor()+0x3c
```

There's a SIGSEGV thrown by `cuCtxSetCurrent` in `CudaMgr::setContext` when unloading a module.

**Pull request merged**